File: app/src/main/java/com/checkin/app/checkin/Cook/Fragment/CookTablesFragment.java
package com.checkin.app.checkin.Cook.Fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.checkin.app.checkin.Cook.Adapter.CookWorkTableAdapter;
import com.checkin.app.checkin.Cook.CookSessionActivity;
import com.checkin.app.checkin.Cook.CookWorkViewModel;
import com.checkin.app.checkin.Cook.Model.CookTableModel;
import com.checkin.app.checkin.Data.Message.MessageModel;
import com.checkin.app.checkin.Data.Message.MessageObjectModel;
import com.checkin.app.checkin.Data.Message.MessageUtils;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.misc.fragments.BaseFragment;
import com.checkin.app.checkin.misc.models.BriefModel;
import com.checkin.app.checkin.R;
import com.checkin.app.checkin.Utility.Utils;

import java.util.Calendar;
import java.util.List;

import butterknife.BindView;

import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_END;
import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_HOST_ASSIGNED;
import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_NEW;
import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_NEW_ORDER;
import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_SWITCH_TABLE;
import static com.checkin.app.checkin.Data.Message.MessageModel.MESSAGE_TYPE.COOK_SESSION_UPDATE_ORDER;

public class CookTablesFragment extends BaseFragment implements CookWorkTableAdapter.CookTableInteraction {

    @BindView(R.id.rv_shop_cook_tables)
    RecyclerView rvShopManagerTable;
    @BindView(R.id.ll_no_orders)
    LinearLayout llNoLiveOrders;

    private CookWorkTableAdapter mAdapter;
    private CookWorkViewModel mViewModel;
    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            MessageModel message = MessageUtils.parseMessage(intent);
            if (message == null) return;

            MessageObjectModel shop = message.getShopDetail();
            if (shop != null && shop.getPk() != mViewModel.getShopPk())
                return;

            BriefModel user;
            switch (message.getType()) {
                case COOK_SESSION_NEW:
                    MessageObjectModel sessionData = message.getSessionDetail();
                    assert sessionData != null;

                    CookTableModel tableModel = new CookTableModel(sessionData.getPk(), null);
                    tableModel.setCreated(Calendar.getInstance().getTime());
                    tableModel.setTable(message.getRawData().getSessionTableName());
                    if (message.getActor().getType() == MessageObjectModel.MESSAGE_OBJECT_TYPE.RESTAURANT_MEMBER) {
                        user = message.getActor().getBriefModel();
                        tableModel.setHost(user);
                    }
                    CookTablesFragment.this.addTable(tableModel);
                    break;
                case COOK_SESSION_NEW_ORDER:
                    CookTablesFragment.this.updateSessionNewOrder(message.getTarget().getPk());
                    break;
                case COOK_SESSION_HOST_ASSIGNED:
                    user = message.getObject().getBriefModel();
                    CookTablesFragment.this.updateSessionHost(message.getTarget().getPk(), user);
                    break;
                case COOK_SESSION_SWITCH_TABLE:
                    mViewModel.fetchActiveTables(mViewModel.getShopPk());
                    break;
                case COOK_SESSION_END:
                    assert message.getSessionDetail() != null;
                    CookTablesFragment.this.removeTable(message.getSessionDetail().getPk());
                    break;
            }
        }
    };

    public CookTablesFragment() {
    }

    public static CookTablesFragment newInstance() {
        return new CookTablesFragment();
    }

    @Override
    protected int getRootLayout() {
        return R.layout.fragment_shop_cook_tables;
    }

    private void updateUi(List<CookTableModel> data) {
        if (data.size() > 0) {
            mAdapter.setRestaurantTableList(data);
            rvShopManagerTable.setVisibility(View.VISIBLE);
            llNoLiveOrders.setVisibility(View.GONE);
        } else {
            rvShopManagerTable.setVisibility(View.GONE);
            llNoLiveOrders.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        mAdapter = new CookWorkTableAdapter(this);
        rvShopManagerTable.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));
        rvShopManagerTable.setAdapter(mAdapter);

        mViewModel = ViewModelProviders.of(requireActivity()).get(CookWorkViewModel.class);
        mViewModel.getActiveTables().observe(this, input -> {
            if (input == null) return;
            if (input.getStatus() == Resource.Status.SUCCESS && input.getData() != null) {
                updateUi(input.getData());
            } else if (input.getStatus() != Resource.Status.LOADING) {
                Utils.toast(requireContext(), input.getMessage());

            }
        });
    }

    // region UI-Update
    private void addTable(CookTableModel tableModel) {
        tableModel.increaseEventCount();
        mViewModel.addRestaurantTable(tableModel);
    }

    private void updateSessionNewOrder(long sessionPk) {
        mViewModel.updateTable(sessionPk, true);
    }

    private void updateSessionHost(long sessionPk, BriefModel user) {
        int pos = mViewModel.getTablePositionWithPk(sessionPk);
        CookTableModel table = mViewModel.getTableWithPosition(pos);
        if (table != null) {
            table.setHost(user);
            mAdapter.updateSession(pos);
        }
    }

    private void removeTable(long sessionPk) {
        mViewModel.updateRemoveTable(sessionPk);
    }
    // endregion

    @Override
    public void onResume() {
        super.onResume();
        MessageModel.MESSAGE_TYPE[] types = new MessageModel.MESSAGE_TYPE[]{
                COOK_SESSION_NEW, COOK_SESSION_NEW_ORDER, COOK_SESSION_UPDATE_ORDER, COOK_SESSION_HOST_ASSIGNED,
                COOK_SESSION_SWITCH_TABLE, COOK_SESSION_END
        };
        MessageUtils.registerLocalReceiver(requireContext(), mReceiver, types);
        mViewModel.updateResults();

    }

    @Override
    public void onPause() {
        super.onPause();
        MessageUtils.unregisterLocalReceiver(requireContext(), mReceiver);
    }

    @Override
    public void onClickTable(CookTableModel tableModel) {
        Intent intent = new Intent(getContext(), CookSessionActivity.class);
        intent.putExtra(CookSessionActivity.KEY_SESSION_PK, tableModel.getPk())
                .putExtra(CookSessionActivity.KEY_SHOP_PK, mViewModel.getShopPk());
        startActivity(intent);

        int pos = mViewModel.getTablePositionWithPk(tableModel.getPk());
        tableModel.resetEventCount();
        mAdapter.updateSession(pos);
    }
}

File: app/src/main/java/com/checkin/app/checkin/cook/fragments/CookTablesFragment.kt
package com.checkin.app.checkin.cook.fragments

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import com.checkin.app.checkin.R
import com.checkin.app.checkin.cook.activities.CookSessionActivity
import com.checkin.app.checkin.cook.adapters.CookWorkTableAdapter
import com.checkin.app.checkin.cook.adapters.CookWorkTableAdapter.CookTableInteraction
import com.checkin.app.checkin.cook.viewmodels.CookWorkViewModel
import com.checkin.app.checkin.data.notifications.MESSAGE_TYPE
import com.checkin.app.checkin.data.notifications.MessageObjectModel
import com.checkin.app.checkin.data.notifications.MessageUtils.parseMessage
import com.checkin.app.checkin.data.notifications.MessageUtils.registerLocalReceiver
import com.checkin.app.checkin.data.notifications.MessageUtils.unregisterLocalReceiver
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.misc.fragments.BaseFragment
import com.checkin.app.checkin.misc.models.BriefModel
import com.checkin.app.checkin.session.activesession.chat.SessionChatModel
import com.checkin.app.checkin.session.models.EventBriefModel
import com.checkin.app.checkin.session.models.RestaurantTableModel
import com.checkin.app.checkin.session.models.TableSessionModel
import com.checkin.app.checkin.utility.Utils

class CookTablesFragment : BaseFragment(), CookTableInteraction {
    override val rootLayout: Int = R.layout.fragment_shop_cook_tables

    @BindView(R.id.rv_shop_cook_tables)
    internal lateinit var rvShopManagerTable: RecyclerView
    @BindView(R.id.ll_no_orders)
    internal lateinit var llNoLiveOrders: LinearLayout

    private val mAdapter: CookWorkTableAdapter = CookWorkTableAdapter(this)
    private val mViewModel: CookWorkViewModel by activityViewModels()

    private val mReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val message = parseMessage(intent)
                    ?: return
            val shop = message.shopDetail
            if (shop != null && shop.pk != mViewModel.shopPk) return
            val user: BriefModel
            when (message.type) {
                MESSAGE_TYPE.COOK_SESSION_NEW -> {
                    val qrPk = message.rawData?.sessionQRId ?: return
                    val tableName = message.rawData.sessionTableName ?: return
                    val actor = message.actor ?: return
                    val eventModel = EventBriefModel.getFromManagerEventModel(message.rawData.sessionEventBrief
                            ?: return).apply {
                        type = SessionChatModel.CHAT_EVENT_TYPE.EVENT_SESSION_CHECKIN
                    }
                    val sessionData = message.sessionDetail ?: return
                    val tableSessionModel = TableSessionModel(sessionData.pk, null, eventModel)
                    val tableModel = RestaurantTableModel(qrPk, tableName, tableSessionModel).apply {
                        addEvent(eventModel)
                        if (actor.type == MessageObjectModel.MESSAGE_OBJECT_TYPE.RESTAURANT_MEMBER) {
                            tableSession?.host = message.actor.briefModel
                        }
                    }
                    addTable(tableModel)
                }
                MESSAGE_TYPE.COOK_SESSION_NEW_ORDER -> updateSessionNewOrder(message.target!!.pk, message.`object`!!.pk)
                MESSAGE_TYPE.COOK_SESSION_HOST_ASSIGNED -> {
                    user = message.`object`!!.briefModel
                    updateSessionHost(message.target!!.pk, user)
                }
                MESSAGE_TYPE.COOK_SESSION_SWITCH_TABLE -> mViewModel.fetchActiveTables(mViewModel.shopPk)
                MESSAGE_TYPE.COOK_SESSION_END -> {
                    removeTable(message.sessionDetail?.pk ?: return)
                }
            }
        }
    }

    private fun updateUi(data: List<RestaurantTableModel>) {
        if (data.isNotEmpty()) {
            mAdapter.setRestaurantTableList(data)
            rvShopManagerTable.visibility = View.VISIBLE
            llNoLiveOrders.visibility = View.GONE
        } else {
            rvShopManagerTable.visibility = View.GONE
            llNoLiveOrders.visibility = View.VISIBLE
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        rvShopManagerTable.apply {
            layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
            rvShopManagerTable.adapter = mAdapter
        }
        mViewModel.activeTables.observe(viewLifecycleOwner, Observer {
            it?.let { input ->
                if (input.data != null) updateUi(input.data)
                else if (input.status !== Resource.Status.LOADING) {
                    Utils.toast(requireContext(), input.message)
                }
            }
        })
    }

    // region UI-Update
    private fun addTable(tableModel: RestaurantTableModel) {
        mViewModel.addRestaurantTable(tableModel)
        tableModel.addToDb()
    }

    private fun updateSessionNewOrder(sessionPk: Long, newOrderPk: Long) {
        mViewModel.updateTable(sessionPk, newOrderPk)
    }

    private fun updateSessionHost(sessionPk: Long, user: BriefModel) {
        val pos = mViewModel.getTablePositionWithPk(sessionPk)
        val table = mViewModel.getTableWithPosition(pos) ?: return
        table.tableSession?.host = user
        mAdapter.updateSession(pos)
    }

    private fun removeTable(sessionPk: Long) {
        mViewModel.updateRemoveTable(sessionPk)
    }

    // endregion
    override fun onResume() {
        super.onResume()
        val types = arrayOf(MESSAGE_TYPE.COOK_SESSION_NEW, MESSAGE_TYPE.COOK_SESSION_NEW_ORDER, MESSAGE_TYPE.COOK_SESSION_UPDATE_ORDER, MESSAGE_TYPE.COOK_SESSION_HOST_ASSIGNED, MESSAGE_TYPE.COOK_SESSION_SWITCH_TABLE, MESSAGE_TYPE.COOK_SESSION_END)
        registerLocalReceiver(requireContext(), mReceiver, *types)
        mViewModel.updateResults()
    }

    override fun onPause() {
        super.onPause()
        unregisterLocalReceiver(requireContext(), mReceiver)
    }

    fun openTableDetails(tableModel: RestaurantTableModel) {
        val intent = Intent(context, CookSessionActivity::class.java)
        val sessionPk = tableModel.sessionPk ?: return
        intent.putExtra(CookSessionActivity.KEY_SESSION_PK, sessionPk)
                .putExtra(CookSessionActivity.KEY_SHOP_PK, mViewModel.shopPk)
        startActivity(intent)
        val pos = mViewModel.getTablePositionWithPk(sessionPk).takeIf { it != -1 } ?: return
        tableModel.resetEvents()
        mAdapter.updateSession(pos)
    }

    override fun onClickTable(tableModel: RestaurantTableModel) = openTableDetails(tableModel)

    companion object {
        fun newInstance(): CookTablesFragment = CookTablesFragment()
    }
}